ukvm/ukvm_hv_openbsd_x86_64.c

```diff
@@ -0,0 +1,254 @@
+            .vrs_gprs[VCPU_REGS_RDI] = X86_BOOT_INFO_BASE,
+            .vrs_crs[VCPU_REGS_CR0] = X86_CR0_INIT,
+            .vrs_crs[VCPU_REGS_CR3] = X86_CR3_INIT,
+            .vrs_crs[VCPU_REGS_CR4] = (X86_CR4_INIT | X86_CR4_VMXE),
+            .vrs_sregs[VCPU_REGS_CS] = sreg_to_vsi(&ukvm_x86_sreg_code),
+            .vrs_sregs[VCPU_REGS_DS] = sreg_to_vsi(&ukvm_x86_sreg_data),
+            .vrs_sregs[VCPU_REGS_ES] = sreg_to_vsi(&ukvm_x86_sreg_data),
+            .vrs_sregs[VCPU_REGS_FS] = sreg_to_vsi(&ukvm_x86_sreg_data),
+            .vrs_sregs[VCPU_REGS_GS] = sreg_to_vsi(&ukvm_x86_sreg_data),
+            .vrs_sregs[VCPU_REGS_SS] = sreg_to_vsi(&ukvm_x86_sreg_data),
+            .vrs_gdtr = { 0x0, X86_GDTR_LIMIT, 0x0, X86_GDT_BASE},
+            .vrs_idtr = { 0x0, 0xFFFF, 0x0, 0x0},
+            .vrs_sregs[VCPU_REGS_LDTR] = sreg_to_vsi(&ukvm_x86_sreg_unusable),
+            .vrs_sregs[VCPU_REGS_TR] = sreg_to_vsi(&ukvm_x86_sreg_tr),
+            .vrs_msrs[VCPU_REGS_EFER] = X86_EFER_INIT,
+            .vrs_msrs[VCPU_REGS_STAR] = 0ULL,
+            .vrs_msrs[VCPU_REGS_LSTAR] = 0ULL,
+            .vrs_msrs[VCPU_REGS_CSTAR] = 0ULL,
+            .vrs_msrs[VCPU_REGS_SFMASK] = 0ULL,
+            .vrs_msrs[VCPU_REGS_KGSBASE] = 0ULL,
+            .vrs_crs[VCPU_REGS_XCR0] = XCR0_X87
+        }
+    };
+
+    ukvm_x86_setup_gdt(hv->mem);
+    ukvm_x86_setup_pagetables(hv->mem, hv->mem_size);
+
+    struct ukvm_boot_info *bi =
+        (struct ukvm_boot_info *)(hv->mem + X86_BOOT_INFO_BASE);
+    bi->mem_size = hv->mem_size;
+    bi->kernel_end = gpa_kend;
+    bi->cmdline = X86_CMDLINE_BASE; 
+    bi->cpu.tsc_freq = get_tsc_freq();
+
+    if (ioctl(hvb->vmd_fd, VMM_IOC_RESETCPU, &vrp) < 0)
+        err(1, "Cannot reset VCPU - exiting.");
+
+    *cmdline = (char *)(hv->mem + X86_CMDLINE_BASE);
+}
+
+void ukvm_hv_vcpu_loop(struct ukvm_hv *hv)
+{
+    struct ukvm_hvb         *hvb = hv->b;
+    struct vm_run_params    *vrp;
+    struct passwd *pw;
+    uid_t uid;
+    gid_t gid;
+
+    vrp = malloc(sizeof(struct vm_run_params));
+    if (vrp == NULL)
+        err(1, "calloc vrp");
+
+    vrp->vrp_exit = malloc(sizeof(union vm_exit));
+    if (vrp == NULL)
+        err(1, "calloc vrp_exit");
+
+    vrp->vrp_vm_id = hvb->vcp_id;
+    vrp->vrp_vcpu_id = hvb->vcpu_id;
+    vrp->vrp_continue = 0;
+
+    if ((pw = getpwnam(VMD_USER)) == NULL)
+        err(1, "can't get _vmd user");
+    uid = pw->pw_uid;
+    gid = pw->pw_gid;
+
+    if (chroot(pw->pw_dir) == -1)
+        err(1, "chroot: %s", pw->pw_dir);
+
+    if (chdir("/") == -1)
+        err(1, "chdir(\"/\")");
+
+    if (setgroups(1, &gid) ||
+        setresgid(gid, gid, gid) ||
+        setresuid(uid, uid, uid))
+        err(1, "unable to revoke privs");
+
+    /*
+     * pledge in the vm processes:
+     * stdio - for malloc and basic I/O including events.
+     * vmm - for the vmm ioctls and operations.
+     */
+    if (pledge("stdio vmm", NULL) == -1)
```

Comment:
I don't get why this doesn't include the block device and the network socket.

Reply:
The block device and network socket are opened earlier, so we already have a handle on them and are allowed access.

Reply:
ah, i see, thanks